File: Springboot/src/main/java/com/playarena/controller/WalletController.java
package com.playarena.controller;

import com.playarena.entity.*;
import com.playarena.repository.*;
import com.playarena.security.JwtUtil;
import org.springframework.web.bind.annotation.*;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/wallet")
@CrossOrigin(origins = "http://localhost:5173")
public class WalletController {

    private final UserRepository userRepository;
    private final NXLWalletRepository walletRepository;
    private final NXLTransactionRepository transactionRepository;
    private final JwtUtil jwtUtil;

    public WalletController(
            UserRepository userRepository,
            NXLWalletRepository walletRepository,
            NXLTransactionRepository transactionRepository,
            JwtUtil jwtUtil
    ) {
        this.userRepository = userRepository;
        this.walletRepository = walletRepository;
        this.transactionRepository = transactionRepository;
        this.jwtUtil = jwtUtil;
    }

    private String getEmail(String emailParam, String authHeader) {
        if (emailParam != null && !emailParam.isEmpty()) {
            return emailParam;
        }
        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            String token = authHeader.substring(7);
            if (jwtUtil.validateToken(token)) {
                return jwtUtil.extractEmail(token);
            }
        }
        return null;
    }

    @GetMapping("/balance")
    public Map<String, Object> getWalletBalance(
            @RequestParam(value = "email", required = false) String emailParam,
            @RequestHeader(value = "Authorization", required = false) String authHeader
    ) {
        String email = getEmail(emailParam, authHeader);
        if (email == null) {
            throw new RuntimeException("Email is required");
        }

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

        NXLWallet wallet = walletRepository.findByUserId(user.getId())
                .orElseGet(() -> {
                    NXLWallet newWallet = new NXLWallet();
                    newWallet.setUserId(user.getId());
                    newWallet.setBalance(0);
                    return walletRepository.save(newWallet);
                });

        return Map.of(
                "userId", user.getId(),
                "email", email,
                "nxlCredits", wallet.getBalance(),
                "walletBalance", wallet.getBalance()
        );
    }

    @GetMapping("/transactions")
    public List<NXLTransaction> getWalletTransactions(
            @RequestParam(value = "email", required = false) String emailParam,
            @RequestHeader(value = "Authorization", required = false) String authHeader
    ) {
        String email = getEmail(emailParam, authHeader);
        if (email == null) {
            throw new RuntimeException("Email is required");
        }

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

        return transactionRepository.findByUserIdOrderByDateDesc(user.getId());
    }

    @PostMapping("/recharge")
    @Transactional
    public Map<String, Object> rechargeWallet(@RequestBody Map<String, Object> payload) {
        String email = (String) payload.get("email");
        int amount = Integer.parseInt(payload.get("amount").toString());

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

        NXLWallet wallet = walletRepository.findByUserId(user.getId())
                .orElseGet(() -> {
                    NXLWallet newWallet = new NXLWallet();
                    newWallet.setUserId(user.getId());
                    newWallet.setBalance(0);
                    return walletRepository.save(newWallet);
                });

        // Conversion rate: 1 INR = 0.1 NXL (i.e. recharge of 1000 INR gives 100 NXL)
        // Let's make it 10% credit earning, matching the preset cards on About page:
        // ₹1000 recharge gives 50 NXL credits. Let's calculate: amount / 20 = NXL credits!
        int creditsToEarn = amount / 20;
        if (creditsToEarn <= 0) {
            creditsToEarn = 1; // Minimum
        }

        wallet.setBalance(wallet.getBalance() + creditsToEarn);
        walletRepository.save(wallet);

        // Update User properties
        user.setWalletBalance(wallet.getBalance());
        user.setCredits(wallet.getBalance());
        userRepository.save(user);

        // Transaction log
        NXLTransaction transaction = new NXLTransaction();
        transaction.setUserId(user.getId());
        transaction.setType("EARNED");
        transaction.setAmount(creditsToEarn);
        transaction.setDescription("Wallet recharge of ₹" + amount);
        transaction.setRefId("RECHARGE-" + System.currentTimeMillis());
        transactionRepository.save(transaction);

        return Map.of(
                "success", true,
                "amount", amount,
                "creditsEarned", creditsToEarn,
                "nxlCredits", wallet.getBalance()
        );
    }

    @PostMapping("/admin/adjust")
    @Transactional
    public Map<String, Object> adminAdjustWallet(@RequestBody Map<String, Object> payload) {
        String email = (String) payload.get("email");
        int amount = Integer.parseInt(payload.get("amount").toString());
        String actionType = (String) payload.get("action"); // ADD or SUBTRACT

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

        NXLWallet wallet = walletRepository.findByUserId(user.getId())
                .orElseGet(() -> {
                    NXLWallet newWallet = new NXLWallet();
                    newWallet.setUserId(user.getId());
                    newWallet.setBalance(0);
                    return walletRepository.save(newWallet);
                });

        if ("SUBTRACT".equalsIgnoreCase(actionType)) {
            if (wallet.getBalance() < amount) {
                throw new RuntimeException("User does not have sufficient NXL balance");
            }
            wallet.setBalance(wallet.getBalance() - amount);
            
            NXLTransaction tx = new NXLTransaction();
            tx.setUserId(user.getId());
            tx.setType("ADMIN_SUB");
            tx.setAmount(amount);
            tx.setDescription("Deducted by administrator adjustment");
            tx.setRefId("ADMIN-" + System.currentTimeMillis());
            transactionRepository.save(tx);
        } else {
            wallet.setBalance(wallet.getBalance() + amount);
            
            NXLTransaction tx = new NXLTransaction();
            tx.setUserId(user.getId());
            tx.setType("ADMIN_ADD");
            tx.setAmount(amount);
            tx.setDescription("Credited by administrator adjustment");
            tx.setRefId("ADMIN-" + System.currentTimeMillis());
            transactionRepository.save(tx);
        }

        walletRepository.save(wallet);

        // Update User entity
        user.setWalletBalance(wallet.getBalance());
        user.setCredits(wallet.getBalance());
        userRepository.save(user);

        return Map.of(
                "success", true,
                "action", actionType,
                "amount", amount,
                "nxlCredits", wallet.getBalance()
        );
    }
}

File: project/controllers/WalletController.php
<?php
require_once __DIR__ . '/../models/User.php';
require_once __DIR__ . '/../models/Wallet.php';
require_once __DIR__ . '/../models/Transaction.php';
require_once __DIR__ . '/../config/JwtUtil.php';

class WalletController {
    private $userModel;
    private $walletModel;
    private $transactionModel;

    public function __construct() {
        $this->userModel = new User();
        $this->walletModel = new Wallet();
        $this->transactionModel = new Transaction();
    }

    private function getEmail($emailParam, $authHeader) {
        if (!empty($emailParam)) {
            return $emailParam;
        }
        if (!empty($authHeader) && strpos($authHeader, "Bearer ") === 0) {
            $token = substr($authHeader, 7);
            if (JwtUtil::validateToken($token)) {
                return JwtUtil::extractEmail($token);
            }
        }
        return null;
    }

    public function getWalletBalance($emailParam, $authHeader) {
        $email = $this->getEmail($emailParam, $authHeader);
        if ($email === null) {
            throw new Exception("Email is required");
        }

        $user = $this->userModel->findByEmail($email);
        if (!$user) {
            throw new Exception("User not found");
        }
        
        if (strtoupper($user['role']) === 'ADMIN') {
            $userIdToUse = 1; // Master admin wallet ID
            // For admins, we just fetch the master wallet balance directly
            $wallet = $this->walletModel->findByUserId($userIdToUse);
            if (!$wallet) {
                $wallet = $this->walletModel->create($userIdToUse, 100000);
            }
            $walletBalance = floatval($wallet['balance']);
            $userCredits = $walletBalance;
        } else {
            $userCredits = floatval($user['credits']);
            $userIdToUse = $user['id'];

            $wallet = $this->walletModel->findByUserId($userIdToUse);
            if (!$wallet) {
                $wallet = $this->walletModel->create($userIdToUse, $userCredits);
            }

            $walletBalance = floatval($wallet['balance']);
            
            // Self-heal out of sync balances for non-admins
            if ($userCredits !== $walletBalance) {
                $correctBalance = max($userCredits, $walletBalance);
                $this->walletModel->updateBalance($userIdToUse, $correctBalance);
                $this->userModel->updateCredits($userIdToUse, $correctBalance);
                $this->userModel->updateWalletBalance($userIdToUse, $correctBalance);
                $walletBalance = $correctBalance;
            }
        }

        return [
            "userId" => $user['id'], // Keep their actual ID in the response for UI
            "email" => $email,
            "nxlCredits" => $walletBalance,
            "walletBalance" => $walletBalance
        ];
    }

    public function getWalletTransactions($emailParam, $authHeader) {
        $email = $this->getEmail($emailParam, $authHeader);
        if ($email === null) {
            throw new Exception("Email is required");
        }

        $user = $this->userModel->findByEmail($email);
        if (!$user) {
            throw new Exception("User not found");
        }

        $userIdToUse = (strtoupper($user['role']) === 'ADMIN') ? 1 : $user['id'];
        return $this->transactionModel->findByUserId($userIdToUse);
    }

    public function rechargeWallet($data) {
        $email = $data['email'] ?? '';
        $amount = floatval($data['amount'] ?? 0);

        $user = $this->userModel->findByEmail($email);
        if (!$user) {
            throw new Exception("User not found");
        }

        $wallet = $this->walletModel->findByUserId($user['id']);
        if (!$wallet) {
            $wallet = $this->walletModel->create($user['id'], 0);
        }

        // Conversion rate: recharge of 100 INR gives 105 NXL credits (amount * 1.05 = NXL)
        $creditsToEarn = intval($amount * 1.05);
        if ($creditsToEarn <= 0) {
            $creditsToEarn = 1;
        }

        // Deduct from master admin wallet (ID 1)
        if ($user['id'] != 1) { 
            $masterUser = $this->userModel->findById(1);
            if ($masterUser && $masterUser['credits'] >= $creditsToEarn) {
                $this->userModel->updateCredits(1, $masterUser['credits'] - $creditsToEarn);
                $this->userModel->updateWalletBalance(1, $masterUser['wallet_balance'] - $creditsToEarn);
                $adminWallet = $this->walletModel->findByUserId(1);
                if ($adminWallet) {
                    $this->walletModel->updateBalance(1, $adminWallet['balance'] - $creditsToEarn);
                }
            } else {
                throw new Exception("Insufficient NXL reserves in Admin Master Wallet!");
            }
        }

        $newBalance = $wallet['balance'] + $creditsToEarn;
        $this->walletModel->updateBalance($user['id'], $newBalance);

        // Update User properties
        $this->userModel->updateWalletBalance($user['id'], $newBalance);
        $this->userModel->updateCredits($user['id'], $newBalance);

        // Log transaction
        $refId = "RECHARGE-" . round(microtime(true) * 1000);
        $this->transactionModel->create($user['id'], "EARNED", $creditsToEarn, "Wallet recharge of ₹" . $amount, $refId);

        return [
            "success" => true,
            "amount" => $amount,
            "creditsEarned" => $creditsToEarn,
            "nxlCredits" => $newBalance
        ];
    }

    public function adminAdjustWallet($data) {
        $email = $data['email'] ?? '';
        $amount = floatval($data['amount'] ?? 0);
        $actionType = $data['action'] ?? 'ADD';

        $user = $this->userModel->findByEmail($email);
        if (!$user) {
            throw new Exception("User not found");
        }

        $wallet = $this->walletModel->findByUserId($user['id']);
        if (!$wallet) {
            $wallet = $this->walletModel->create($user['id'], 0);
        }

        $newBalance = $wallet['balance'];
        $refId = "ADMIN-" . round(microtime(true) * 1000);

        if (strcasecmp($actionType, "SUBTRACT") === 0) {
            if ($wallet['balance'] < $amount) {
                throw new Exception("User does not have sufficient NXL balance");
            }
            $newBalance -= $amount;
            $this->transactionModel->create($user['id'], "ADMIN_SUB", $amount, "Deducted by administrator adjustment", $refId);
            
            // Refund to master admin wallet
            $this->userModel->updateCredits(1, $this->userModel->findById(1)['credits'] + $amount);
            $this->userModel->updateWalletBalance(1, $this->userModel->findById(1)['wallet_balance'] + $amount);
            $masterWallet = $this->walletModel->findByUserId(1);
            if ($masterWallet) $this->walletModel->updateBalance(1, $masterWallet['balance'] + $amount);
        } else {
            $newBalance += $amount;
            $this->transactionModel->create($user['id'], "ADMIN_ADD", $amount, "Credited by administrator adjustment", $refId);
            
            // Deduct from master admin wallet
            $masterUser = $this->userModel->findById(1);
            if ($masterUser && $masterUser['credits'] >= $amount) {
                $this->userModel->updateCredits(1, $masterUser['credits'] - $amount);
                $this->userModel->updateWalletBalance(1, $masterUser['wallet_balance'] - $amount);
                $masterWallet = $this->walletModel->findByUserId(1);
                if ($masterWallet) $this->walletModel->updateBalance(1, $masterWallet['balance'] - $amount);
            }
        }

        $this->walletModel->updateBalance($user['id'], $newBalance);

        // Update User properties
        $this->userModel->updateWalletBalance($user['id'], $newBalance);
        $this->userModel->updateCredits($user['id'], $newBalance);

        return [
            "success" => true,
            "action" => $actionType,
            "amount" => $amount,
            "nxlCredits" => $newBalance
        ];
    }

    public function getAllTransactions() {
        try {
            $transactions = $this->transactionModel->findAllWithUsers(100);
            return [
                "success" => true,
                "transactions" => $transactions
            ];
        } catch (Exception $e) {
            return [
                "success" => false,
                "message" => "An error occurred: " . $e->getMessage()
            ];
        }
    }
}
